File: flutter/lib/core/model/overlay.dart
import 'dart:math' as math;

import 'ids.dart';

enum OverlayAnimation {
  none('None'),
  fade('Fade'),
  slideUp('Slide up'),
  slideDown('Slide down'),
  pop('Pop'),
  typewriter('Typewriter');

  const OverlayAnimation(this.label);

  final String label;
}

/// Placement of an overlay, as fractions of the canvas.
///
/// Fractions rather than pixels so a project opens correctly at any export
/// resolution: text positioned on a 1080p preview lands in the same place in a
/// 4K render without being rescaled by hand.
class OverlayTransform {
  const OverlayTransform({
    this.x = 0.5,
    this.y = 0.5,
    this.scale = 1,
    this.rotationDegrees = 0,
    this.opacity = 1,
  });

  final double x;
  final double y;
  final double scale;
  final double rotationDegrees;
  final double opacity;

  OverlayTransform copyWith({
    double? x,
    double? y,
    double? scale,
    double? rotationDegrees,
    double? opacity,
  }) =>
      OverlayTransform(
        x: x ?? this.x,
        y: y ?? this.y,
        scale: scale ?? this.scale,
        rotationDegrees: rotationDegrees ?? this.rotationDegrees,
        opacity: opacity ?? this.opacity,
      );

  Map<String, dynamic> toJson() => {
        'x': x,
        'y': y,
        'scale': scale,
        'rotationDegrees': rotationDegrees,
        'opacity': opacity,
      };

  static OverlayTransform fromJson(Map<String, dynamic> json) =>
      OverlayTransform(
        x: (json['x'] as num?)?.toDouble() ?? 0.5,
        y: (json['y'] as num?)?.toDouble() ?? 0.5,
        scale: (json['scale'] as num?)?.toDouble() ?? 1,
        rotationDegrees: (json['rotationDegrees'] as num?)?.toDouble() ?? 0,
        opacity: (json['opacity'] as num?)?.toDouble() ?? 1,
      );
}

abstract class Overlay {
  const Overlay();

  String get id;
  int get startUs;
  int get endUs;
  OverlayTransform get transform;
  OverlayAnimation get animationIn;
  OverlayAnimation get animationOut;
  int get animationDurationUs;

  int get durationUs => math.max(0, endUs - startUs);

  bool isActiveAt(int timeUs) => timeUs >= startUs && timeUs < endUs;

  /// Progress through the entry animation; 1 once it has finished.
  double enterProgress(int timeUs) {
    if (animationDurationUs <= 0 || animationIn == OverlayAnimation.none) {
      return 1;
    }
    final elapsed = timeUs - startUs;
    if (elapsed >= animationDurationUs) return 1;
    return (elapsed / animationDurationUs).clamp(0.0, 1.0);
  }

  /// Progress through the exit animation; 1 while it has not begun.
  double exitProgress(int timeUs) {
    if (animationDurationUs <= 0 || animationOut == OverlayAnimation.none) {
      return 1;
    }
    final remaining = endUs - timeUs;
    if (remaining >= animationDurationUs) return 1;
    return (remaining / animationDurationUs).clamp(0.0, 1.0);
  }

  String get trackLabel;

  Overlay withTiming({int? startUs, int? endUs});

  Overlay withTransform(OverlayTransform transform);

  Map<String, dynamic> toJson();
}

class TextStyleSpec {
  const TextStyleSpec({
    this.fontSize = 0.07, // fraction of canvas height
    this.color = 0xFFFFFFFF,
    this.bold = true,
    this.italic = false,
    this.align = TextAlignSpec.center,
    this.outlineWidth = 0.14, // fraction of font size
    this.outlineColor = 0xFF000000,
    this.plateColor = 0x00000000,
    this.letterSpacing = 0,
    this.lineHeight = 1.15,
  });

  final double fontSize;
  final int color;
  final bool bold;
  final bool italic;
  final TextAlignSpec align;
  final double outlineWidth;
  final int outlineColor;
  final int plateColor;
  final double letterSpacing;
  final double lineHeight;

  TextStyleSpec copyWith({
    double? fontSize,
    int? color,
    bool? bold,
    bool? italic,
    TextAlignSpec? align,
    double? outlineWidth,
    int? outlineColor,
    int? plateColor,
    double? letterSpacing,
    double? lineHeight,
  }) =>
      TextStyleSpec(
        fontSize: fontSize ?? this.fontSize,
        color: color ?? this.color,
        bold: bold ?? this.bold,
        italic: italic ?? this.italic,
        align: align ?? this.align,
        outlineWidth: outlineWidth ?? this.outlineWidth,
        outlineColor: outlineColor ?? this.outlineColor,
        plateColor: plateColor ?? this.plateColor,
        letterSpacing: letterSpacing ?? this.letterSpacing,
        lineHeight: lineHeight ?? this.lineHeight,
      );

  Map<String, dynamic> toJson() => {
        'fontSize': fontSize,
        'color': color,
        'bold': bold,
        'italic': italic,
        'align': align.name,
        'outlineWidth': outlineWidth,
        'outlineColor': outlineColor,
        'plateColor': plateColor,
        'letterSpacing': letterSpacing,
        'lineHeight': lineHeight,
      };

  static TextStyleSpec fromJson(Map<String, dynamic> json) => TextStyleSpec(
        fontSize: (json['fontSize'] as num?)?.toDouble() ?? 0.07,
        color: (json['color'] as num?)?.toInt() ?? 0xFFFFFFFF,
        bold: json['bold'] as bool? ?? true,
        italic: json['italic'] as bool? ?? false,
        align: TextAlignSpec.values.firstWhere(
          (a) => a.name == json['align'],
          orElse: () => TextAlignSpec.center,
        ),
        outlineWidth: (json['outlineWidth'] as num?)?.toDouble() ?? 0.14,
        outlineColor: (json['outlineColor'] as num?)?.toInt() ?? 0xFF000000,
        plateColor: (json['plateColor'] as num?)?.toInt() ?? 0x00000000,
        letterSpacing: (json['letterSpacing'] as num?)?.toDouble() ?? 0,
        lineHeight: (json['lineHeight'] as num?)?.toDouble() ?? 1.15,
      );
}

enum TextAlignSpec { left, center, right }

class TextOverlay extends Overlay {
  TextOverlay({
    String? id,
    this.text = 'Your text',
    this.style = const TextStyleSpec(),
    this.startUs = 0,
    this.endUs = 3000000,
    this.transform = const OverlayTransform(),
    this.animationIn = OverlayAnimation.fade,
    this.animationOut = OverlayAnimation.fade,
    this.animationDurationUs = 300000,
  }) : id = id ?? newId();

  @override
  final String id;
  final String text;
  final TextStyleSpec style;
  @override
  final int startUs;
  @override
  final int endUs;
  @override
  final OverlayTransform transform;
  @override
  final OverlayAnimation animationIn;
  @override
  final OverlayAnimation animationOut;
  @override
  final int animationDurationUs;

  @override
  String get trackLabel => text.trim().isEmpty ? 'Text' : text.trim();

  TextOverlay copyWith({
    String? text,
    TextStyleSpec? style,
    int? startUs,
    int? endUs,
    OverlayTransform? transform,
    OverlayAnimation? animationIn,
    OverlayAnimation? animationOut,
    int? animationDurationUs,
  }) =>
      TextOverlay(
        id: id,
        text: text ?? this.text,
        style: style ?? this.style,
        startUs: startUs ?? this.startUs,
        endUs: endUs ?? this.endUs,
        transform: transform ?? this.transform,
        animationIn: animationIn ?? this.animationIn,
        animationOut: animationOut ?? this.animationOut,
        animationDurationUs: animationDurationUs ?? this.animationDurationUs,
      );

  @override
  Overlay withTiming({int? startUs, int? endUs}) =>
      copyWith(startUs: startUs, endUs: endUs);

  @override
  Overlay withTransform(OverlayTransform transform) =>
      copyWith(transform: transform);

  @override
  Map<String, dynamic> toJson() => {
        'kind': 'text',
        'id': id,
        'text': text,
        'style': style.toJson(),
        'startUs': startUs,
        'endUs': endUs,
        'transform': transform.toJson(),
        'animationIn': animationIn.name,
        'animationOut': animationOut.name,
        'animationDurationUs': animationDurationUs,
      };
}

class StickerOverlay extends Overlay {
  StickerOverlay({
    String? id,
    this.emoji = '⭐',
    this.sizeFraction = 0.2,
    this.startUs = 0,
    this.endUs = 3000000,
    this.transform = const OverlayTransform(),
    this.animationIn = OverlayAnimation.pop,
    this.animationOut = OverlayAnimation.fade,
    this.animationDurationUs = 300000,
  }) : id = id ?? newId();

  @override
  final String id;
  final String emoji;

  /// Size as a fraction of canvas height.
  final double sizeFraction;
  @override
  final int startUs;
  @override
  final int endUs;
  @override
  final OverlayTransform transform;
  @override
  final OverlayAnimation animationIn;
  @override
  final OverlayAnimation animationOut;
  @override
  final int animationDurationUs;

  @override
  String get trackLabel => emoji;

  StickerOverlay copyWith({
    String? emoji,
    double? sizeFraction,
    int? startUs,
    int? endUs,
    OverlayTransform? transform,
    OverlayAnimation? animationIn,
    OverlayAnimation? animationOut,
    int? animationDurationUs,
  }) =>
      StickerOverlay(
        id: id,
        emoji: emoji ?? this.emoji,
        sizeFraction: sizeFraction ?? this.sizeFraction,
        startUs: startUs ?? this.startUs,
        endUs: endUs ?? this.endUs,
        transform: transform ?? this.transform,
        animationIn: animationIn ?? this.animationIn,
        animationOut: animationOut ?? this.animationOut,
        animationDurationUs: animationDurationUs ?? this.animationDurationUs,
      );

  @override
  Overlay withTiming({int? startUs, int? endUs}) =>
      copyWith(startUs: startUs, endUs: endUs);

  @override
  Overlay withTransform(OverlayTransform transform) =>
      copyWith(transform: transform);

  @override
  Map<String, dynamic> toJson() => {
        'kind': 'sticker',
        'id': id,
        'emoji': emoji,
        'sizeFraction': sizeFraction,
        'startUs': startUs,
        'endUs': endUs,
        'transform': transform.toJson(),
        'animationIn': animationIn.name,
        'animationOut': animationOut.name,
        'animationDurationUs': animationDurationUs,
      };
}

OverlayAnimation _animation(Object? name, OverlayAnimation fallback) =>
    OverlayAnimation.values.firstWhere(
      (a) => a.name == name,
      orElse: () => fallback,
    );

Overlay? overlayFromJson(Map<String, dynamic> json) {
  final transform = json['transform'] is Map
      ? OverlayTransform.fromJson(
          Map<String, dynamic>.from(json['transform'] as Map))
      : const OverlayTransform();
  final startUs = (json['startUs'] as num?)?.toInt() ?? 0;
  final endUs = (json['endUs'] as num?)?.toInt() ?? 3000000;
  final animationDurationUs =
      (json['animationDurationUs'] as num?)?.toInt() ?? 300000;

  switch (json['kind']) {
    case 'text':
      return TextOverlay(
        id: json['id'] as String?,
        text: json['text'] as String? ?? '',
        style: json['style'] is Map
            ? TextStyleSpec.fromJson(
                Map<String, dynamic>.from(json['style'] as Map))
            : const TextStyleSpec(),
        startUs: startUs,
        endUs: endUs,
        transform: transform,
        animationIn: _animation(json['animationIn'], OverlayAnimation.fade),
        animationOut: _animation(json['animationOut'], OverlayAnimation.fade),
        animationDurationUs: animationDurationUs,
      );
    case 'sticker':
      return StickerOverlay(
        id: json['id'] as String?,
        emoji: json['emoji'] as String? ?? '⭐',
        sizeFraction: (json['sizeFraction'] as num?)?.toDouble() ?? 0.2,
        startUs: startUs,
        endUs: endUs,
        transform: transform,
        animationIn: _animation(json['animationIn'], OverlayAnimation.pop),
        animationOut: _animation(json['animationOut'], OverlayAnimation.fade),
        animationDurationUs: animationDurationUs,
      );
    default:
      return null;
  }
}

File: android/app/src/main/java/com/vixel/studio/core/model/Overlay.kt
package com.vixel.studio.core.model

import java.util.UUID
import kotlin.math.max

/** How an overlay enters or leaves. */
enum class OverlayAnimation(val label: String) {
    NONE("None"),
    FADE("Fade"),
    SLIDE_UP("Slide up"),
    SLIDE_DOWN("Slide down"),
    SLIDE_LEFT("Slide left"),
    SLIDE_RIGHT("Slide right"),
    POP("Pop"),
    ZOOM("Zoom"),
}

/**
 * Placement on the canvas.
 *
 * Position is a fraction of the canvas so an overlay lands in the same place
 * whether the project is previewed at 540p or exported at 2160p, and [scale]
 * multiplies a size that is itself expressed relative to canvas height.
 */
data class OverlayTransform(
    /** 0 = left edge, 1 = right edge. */
    val x: Float = 0.5f,
    /** 0 = top edge, 1 = bottom edge. */
    val y: Float = 0.5f,
    val scale: Float = 1f,
    val rotationDegrees: Float = 0f,
    val opacity: Float = 1f,
)

sealed interface Overlay {
    val id: String
    val startUs: Long
    val endUs: Long
    val transform: OverlayTransform
    val animationIn: OverlayAnimation
    val animationOut: OverlayAnimation
    val animationDurationUs: Long

    val durationUs: Long get() = max(0L, endUs - startUs)

    fun isActiveAt(timeUs: Long): Boolean = timeUs >= startUs && timeUs < endUs

    /** Progress through the entry animation, 1 once it has finished. */
    fun enterProgress(timeUs: Long): Float {
        if (animationDurationUs <= 0L || animationIn == OverlayAnimation.NONE) return 1f
        val elapsed = timeUs - startUs
        if (elapsed >= animationDurationUs) return 1f
        return (elapsed.toFloat() / animationDurationUs).coerceIn(0f, 1f)
    }

    /** Progress through the exit animation, 1 while it has not started. */
    fun exitProgress(timeUs: Long): Float {
        if (animationDurationUs <= 0L || animationOut == OverlayAnimation.NONE) return 1f
        val remaining = endUs - timeUs
        if (remaining >= animationDurationUs) return 1f
        return (remaining.toFloat() / animationDurationUs).coerceIn(0f, 1f)
    }

    fun withTiming(startUs: Long, endUs: Long): Overlay
    fun withTransform(transform: OverlayTransform): Overlay
}

// ------------------------------------------------------------------- text

enum class TextFont(val label: String) {
    SANS("Sans"),
    SANS_CONDENSED("Condensed"),
    SERIF("Serif"),
    MONOSPACE("Mono"),
    CURSIVE("Cursive"),
}

enum class TextAlignment(val label: String) { START("Left"), CENTER("Center"), END("Right") }

data class TextStyle(
    val font: TextFont = TextFont.SANS,
    val bold: Boolean = true,
    val italic: Boolean = false,
    /** Cap height as a fraction of canvas height, so it scales with output. */
    val sizeFraction: Float = 0.07f,
    val color: Int = 0xFFFFFFFF.toInt(),
    val alignment: TextAlignment = TextAlignment.CENTER,
    val letterSpacing: Float = 0f,
    val lineSpacing: Float = 1.1f,

    /** Outline. 0 disables it. */
    val strokeWidth: Float = 0f,
    val strokeColor: Int = 0xFF000000.toInt(),

    /** Drop shadow. 0 radius disables it. */
    val shadowRadius: Float = 0f,
    val shadowDx: Float = 0f,
    val shadowDy: Float = 0.12f,
    val shadowColor: Int = 0xB3000000.toInt(),

    /** Plate behind the text. Fully transparent disables it. */
    val backgroundColor: Int = 0x00000000,
    val backgroundPadding: Float = 0.35f,
    val backgroundRadius: Float = 0.25f,
)

data class TextOverlay(
    override val id: String = UUID.randomUUID().toString(),
    val text: String = "Your text",
    val style: TextStyle = TextStyle(),
    override val startUs: Long = 0L,
    override val endUs: Long = 3_000_000L,
    override val transform: OverlayTransform = OverlayTransform(),
    override val animationIn: OverlayAnimation = OverlayAnimation.FADE,
    override val animationOut: OverlayAnimation = OverlayAnimation.FADE,
    override val animationDurationUs: Long = 400_000L,
) : Overlay {
    override fun withTiming(startUs: Long, endUs: Long): Overlay =
        copy(startUs = startUs, endUs = endUs)

    override fun withTransform(transform: OverlayTransform): Overlay =
        copy(transform = transform)
}

// ---------------------------------------------------------------- stickers

/**
 * Sticker artwork.
 *
 * Emoji come from the system font and shapes are drawn as paths, so the
 * library costs nothing in APK size and renders crisply at any resolution.
 */
sealed interface StickerArt {
    data class Emoji(val glyph: String) : StickerArt
    data class Shape(val shape: StickerShape, val color: Int) : StickerArt
}

enum class StickerShape(val label: String) {
    CIRCLE("Circle"),
    SQUARE("Square"),
    TRIANGLE("Triangle"),
    STAR("Star"),
    HEART("Heart"),
    ARROW("Arrow"),
    SPEECH_BUBBLE("Bubble"),
    BURST("Burst"),
}

data class StickerOverlay(
    override val id: String = UUID.randomUUID().toString(),
    val art: StickerArt = StickerArt.Emoji("⭐"),
    /** Size as a fraction of canvas height. */
    val sizeFraction: Float = 0.2f,
    override val startUs: Long = 0L,
    override val endUs: Long = 3_000_000L,
    override val transform: OverlayTransform = OverlayTransform(),
    override val animationIn: OverlayAnimation = OverlayAnimation.POP,
    override val animationOut: OverlayAnimation = OverlayAnimation.FADE,
    override val animationDurationUs: Long = 350_000L,
) : Overlay {
    override fun withTiming(startUs: Long, endUs: Long): Overlay =
        copy(startUs = startUs, endUs = endUs)

    override fun withTransform(transform: OverlayTransform): Overlay =
        copy(transform = transform)
}

/** Sticker choices offered in the picker. */
object StickerLibrary {

    val emoji: List<String> = listOf(
        "⭐", "❤️", "🔥", "✨", "👍", "😂",
        "😍", "🤩", "😎", "🎉", "💯", "👀",
        "🎵", "📸", "🌈", "⚡", "🌙", "☀️",
        "🌿", "🌺", "🍕", "☕", "👋", "🙌",
    )

    val shapeColors: List<Int> = listOf(
        0xFFFFFFFF.toInt(), 0xFF000000.toInt(), 0xFF6FE3C9.toInt(), 0xFFB08CFF.toInt(),
        0xFF5AA9FF.toInt(), 0xFFFFC46B.toInt(), 0xFFFF7A90.toInt(),
    )
}
